In [1]:
"""
PYTORCH НЕГІЗДЕРІ: ТЕНЗОРЛАР, АВТОМАТТЫ ДИФФЕРЕНЦИАЦИЯ ЖӘНЕ БІРІНШІ НЕЙРОНДЫҚ ЖЕЛІ
Google Colab-та 5 тапсырманы жеке орындауға арналған код.
"""

# ============================
# 1-ТАПСЫРМА: 3x3 тензорлар және матрицалық көбейту
# ============================
import torch

# Қайта іске қосқанда бірдей нәтижеге жуық болу үшін random seed орнатамыз
torch.manual_seed(42)

# Екі кездейсоқ 3x3 тензор құрамыз
A = torch.randn(3, 3)
B = torch.randn(3, 3)

# Матрицалық көбейтінді (екі түрлі жазылуы да дұрыс)
C = torch.matmul(A, B)   # немесе C = A @ B

print("1-тапсырма:")
print("A тензоры:\n", A)
print("B тензоры:\n", B)
print("A @ B нәтижесі:\n", C)


# ============================
# 2-ТАПСЫРМА: Autograd арқылы градиент есептеу
# ============================
# requires_grad=True дегеніміз: осы тензор бойынша туындыны (градиентті) есептеу керек
x = torch.tensor(2.0, requires_grad=True)

# Функция: y = x^2 + 3x + 1
y = x**2 + 3*x + 1

# Кері таралу: dy/dx градиентін есептейді
y.backward()

print("\n2-тапсырма:")
print("x мәні:", x.item())
print("y мәні:", y.item())
print("x бойынша градиент (dy/dx):", x.grad.item())  # x=2 үшін dy/dx = 2x+3 = 7


# ============================
# 3-ТАПСЫРМА: Жасанды дерек және қарапайым модель құру
# ============================
# 100 жол, 1 баған кіріс
X = torch.randn(100, 1)

# Аздаған шу (noise) қосамыз
noise = 0.1 * torch.randn(100, 1)

# Нақты тәуелділік: y = 3X + 2 + noise
y_data = 3 * X + 2 + noise

# Ең қарапайым нейрондық желі: бір ғана Linear қабаты (1 кіріс -> 1 шығыс)
model = torch.nn.Linear(1, 1)

print("\n3-тапсырма:")
print("X shape:", X.shape)
print("y shape:", y_data.shape)
print("Құрылған модель:", model)


# ============================
# 4-ТАПСЫРМА: Loss және Optimizer таңдау
# ============================
criterion = torch.nn.MSELoss()                 # Орташа квадраттық қателік
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)  # SGD, learning rate = 0.01

print("\n4-тапсырма:")
print("Loss функциясы:", criterion)
print("Optimizer:", optimizer)


# ============================
# 5-ТАПСЫРМА: 100 эпоха оқыту циклі (5 негізгі қадам)
# ============================
num_epochs = 100

for epoch in range(1, num_epochs + 1):
    # 1) Болжам алу (forward pass)
    y_pred = model(X)

    # 2) Қателікті есептеу
    loss = criterion(y_pred, y_data)

    # 3) Ескі градиенттерді нөлдеу
    optimizer.zero_grad()

    # 4) Кері таралу (градиенттерді есептеу)
    loss.backward()

    # 5) Салмақтарды жаңарту
    optimizer.step()

    # Әр 10 эпоха сайын loss басып шығару
    if epoch % 10 == 0:
        print(f"Epoch [{epoch:3d}/{num_epochs}] - Loss: {loss.item():.6f}")


# Оқыту соңындағы параметрлерді қарау (w және b шамамен 3 және 2-ге жақындауы керек)
learned_w = model.weight.item()
learned_b = model.bias.item()

print("\nОқыту аяқталды:")
print(f"Үйренген weight (w): {learned_w:.4f}")
print(f"Үйренген bias   (b): {learned_b:.4f}")


1-тапсырма:
A тензоры:
 tensor([[ 0.3367,  0.1288,  0.2345],
        [ 0.2303, -1.1229, -0.1863],
        [ 2.2082, -0.6380,  0.4617]])
B тензоры:
 tensor([[ 0.2674,  0.5349,  0.8094],
        [ 1.1103, -1.6898, -0.9890],
        [ 0.9580,  1.3221,  0.8172]])
A @ B нәтижесі:
 tensor([[ 0.4576,  0.2724,  0.3367],
        [-1.3636,  1.7743,  1.1446],
        [ 0.3243,  2.8696,  2.7954]])

2-тапсырма:
x мәні: 2.0
y мәні: 11.0
x бойынша градиент (dy/dx): 7.0

3-тапсырма:
X shape: torch.Size([100, 1])
y shape: torch.Size([100, 1])
Құрылған модель: Linear(in_features=1, out_features=1, bias=True)

4-тапсырма:
Loss функциясы: MSELoss()
Optimizer: SGD (
Parameter Group 0
    dampening: 0
    differentiable: False
    foreach: None
    fused: None
    lr: 0.01
    maximize: False
    momentum: 0
    nesterov: False
    weight_decay: 0
)
Epoch [ 10/100] - Loss: 12.840488
Epoch [ 20/100] - Loss: 8.537443
Epoch [ 30/100] - Loss: 5.683264
Epoch [ 40/100] - Loss: 3.788373
Epoch [ 50/100] - Loss: 2.5



**1-Тапсырма: 3x3 тензорлар және матрицалық көбейту**
- Екі кездейсоқ 3x3 тензор (A және B) сәтті жасалды.
- Олардың матрицалық көбейтіндісі (A @ B) дұрыс есептелді, бұл тензорлық операциялардың PyTorch-та қалай орындалатынын көрсетеді.

**2-Тапсырма: Autograd арқылы градиент есептеу**
- `x = 2.0` мәні үшін `y = x^2 + 3x + 1` функциясы анықталды.
- PyTorch-тың `autograd` механизмі арқылы `dy/dx` градиенті есептелді. Нәтиже `7.0` болды, бұл аналитикалық есептеуге (2x + 3 = 2*2 + 3 = 7) сәйкес келеді. Бұл autograd-тың автоматты дифференциацияны қалай орындайтынын дәлелдейді.

**3, 4 және 5-Тапсырмалар: Нейрондық желіні оқыту**
- **Деректер:** `y = 3X + 2 + noise` формуласы бойынша 100 дерек нүктесі (X, y_data) генерацияланды, мұнда X - кіріс, y_data - шығыс, ал noise - аздаған шу.
- **Модель:** Бір кіріс және бір шығыс нейроны бар өте қарапайым сызықтық модель (`torch.nn.Linear(1, 1)`) құрылды.
- **Loss функциясы мен Оптимизатор:** Модельдің қателігін өлшеу үшін орташа квадраттық қателік (`MSELoss`) және салмақтарды жаңарту үшін Стохастикалық Градиенттік Түсу (`SGD`) оптимизаторы таңдалды.
- **Оқыту:** Модель 100 эпоха бойы оқытылды. Әр эпохада: болжам жасалды, қателік есептелді, градиенттер нөлденді, кері таралу арқылы градиенттер есептелді және оптимизатор салмақтарды жаңартты. Оқыту барысында loss (қателік) мәні `12.84`-тен `0.34`-ке дейін айтарлықтай төмендеді, бұл модельдің деректердегі үлгіні үйреніп жатқанын көрсетеді.
- **Нәтиже:** Оқыту соңында үйренген салмақ (`weight` немесе `w`) **2.5015** және `bias` (`b`) **1.6954** болды. Бұл бастапқы деректерді генерациялаудағы нақты мәндерге (`w=3, b=2`) өте жақын, бұл қарапайым желінің оқыту арқылы деректердегі негізгі байланысты сәтті үйренгенін білдіреді.